### DISCUSSION TITLE
How can I reference the data in a node.js project and react based project

### DISCUSSION THREAD

**Discussion by jetro4u** [OAI/Arazzo-Specification#216] [Q&A] [answered]
I will like to find out how I can reference the data in a node.js project and react based project. Also how to use the workflow in a project

**Comment by kevinduffey** [comment]
I am not sure if I understand what you are asking. I'll give it a shot though. I assume you are referring to the execution of a given workflow as if it were code, such that whatever is returned in the end step, say for example a workflow that buys a pet from a pet store, the result might be a payload (response body) that provides details about the transaction?

**Reply by jetro4u** [reply]
That is one part of what I want to know. 

The other part, I am building an isomorphic openapi adapter that can work both for client and server. I am creating some methods to parse the document schema specs. 

In openapi.json I can easily create all part of the document object as json especially the general one. And the response ai get back from server is also in the same shape. 

For Arazzo, I prefer to use the json format which I think is different from the normal json object. So I am asking if I can use the JSON template format to create all sections of the arazzo.json document including info, documents, workflow and others so that I can modify my parser function. Then the response from a workflow, will it be also in the json format or yaml. I need to know whether I also need to create a parser function for the response. 

I also intend to build or rather update my existing CLI command library to possibly use the workflow automation as I currently used some custom build tools that rely on nodejs.

**Reply by jetro4u** [reply]
I also read this statement "In order to preserve the ability to round-trip between YAML and JSON formats, YAML version 1.2 is RECOMMENDED along with some additional constraints:

Tags MUST be limited to those allowed by the JSON Schema ruleset.

Keys used in YAML maps MUST be limited to a scalar string, as defined by the YAML Failsafe schema ruleset.
"

I'll assume that for best result, YAML 1.2 should be used. 

Can I reference another openapi.json document or other external url?

**Reply by jetro4u** [reply]
I need clarification on this statement "In a multi-document description, the document containing the Arazzo Specification Object is known as the entry Arazzo document."

In the document description, the Arazzo specification object include REQUIRED field. So one can conclude that it should be included in all Arazzo documents. So how will the system differentiate what the entry document is if I define the specification in all of them. Or can I define a file without the specification and reference it from an Arazzo file?

**Reply by handrews** [reply]
@jetro4u 
> I need clarification on this statement "In a multi-document description, the document containing the Arazzo Specification Object is known as the entry Arazzo document."
> 
> In the document description, the Arazzo specification object include REQUIRED field. So one can conclude that it should be included in all Arazzo documents. So how will the system differentiate what the entry document is if I define the specification in all of them. Or can I define a file without the specification and reference it from an Arazzo file?

@kevinduffey @frankkilcommins we tweaked the analogous language in OAS 3.0.4 and OAS 3.1.1 for this reason.  We also did some clarifications of the YAML-related text. See:

* OAI/OpenAPI-Specification#3820
* OAI/OpenAPI-Specification#3861

You might want to adopt something similar in Arazzo 1.0.1.

**Reply by kevinduffey** [reply]
@jetro4u the entry document would be the one that has the imports of other documents. So whatever tool is used to parse/do something would use the Arazzo object (at the top) of the first document loaded.. imported workflow docs would be more akin to inlined workflows where the Arazzo object would likely be ignored (other than perhaps to match version info and/or details of imported docs). In the entry document, you would not utilize the header of imported docs, you would be referencing workflows, steps and components within those docs. 

As @handrews eluded to, and shared links, the Complete Document wording would indicate everything is contained in the one document which I would assume means no imports of other Arazzo documents. Though in the case of Arazzo you likely have OpenAPI documents you import so not entirely sure if "Complete Document" definition would still fit this as it does OpenAPI.

**Reply by jetro4u** [reply]
I'm building an Isomorphic Adapter that also parse the yaml and json document. I guess I'll need to design my project such that only the entry document reference other documents. Given the short time I have it'll be easier to also wrap up the library and expand the functionality later.

**Reply by handrews** [reply, marked as answer]
@kevinduffey actually "complete document" does not mean it's self-contained, just that it's not a fragment (naming is hard!)  A fragment would be something like a document containing _only_ a Path Item Object.  This sort of thing was typical in 2.0 and is still quite common in 3.0.  It causes a lot of problems in OAS 3.1, which is explained much more clearly in the forthcoming 3.1.1.  Although 3.1 makes it more beneficial to have standalone JSON Schema documents, which _technically_ are fragments in OAS terms.  Hopefully 3.1.1 makes more sense in this area as it has been a challenge for a lot of tooling vendors.

So the thing here is that you can have multiple complete documents (at least in OAS) that are connected with references, in which case you have to designate one of them as the entry document somehow (even if it's just "whatever is passed to the tool first is the entry document").  On the other hand, if you only have one complete document and all other documents are fragments, then you can safely assume the complete document is the entry document.  I think that's the scenario everyone had in mind with the original entry document wording (even though that scenario is problematic in OAS 3.1).

**Reply by jetro4u** [reply]
Just curios, is there existing interface that define the Arazzo specs like the OpenAPIV3_1
